_extensions/custom_roles.py

```diff
@@ -19,12 +25,16 @@ def setup(app):
 
 def autolink(pattern):
     def role(name, rawtext, text, lineno, inliner, options={}, content=[]):
+        """Combine literal + reference (unless the text is prefixed by a !)."""
         if " " in text:
             url_text = urllib.parse.quote(f"{text}")
```

Comment:
This is already a string, so the f-string call is unneeded
```suggestion
            url_text = urllib.parse.quote(text)
```